chore(ci): introduce workflows for release automation

achou11 opened the pull request:
Closes #900 

Ports over the workflows originally implemented in https://github.com/digidem/release-automations and subsequently improved and tested in https://github.com/digidem/release-automations-expo. Documentation about what each workflow does is detailed in the former (although it's probably slightly outdated).

I've already added a `EAS_PROJECT_URL` variable for the repo. Looks like our signing keys for the rc and production profiles already exist on EAS.

I have also ported over [rulesets](https://github.com/digidem/release-automations-expo/settings/rules) from https://github.com/digidem/release-automations-expo that are meant to prevent issues during the release process. They are currently NOT enabled but should be once we're on a fresh release cycle (currently in the process of one and I don't want to disrupt it). A brief overview of what each ruleset does:

1. [Default](https://github.com/digidem/comapeo-mobile/settings/rules/3635880)
    
    - Prevents force pushing or deletion of the `develop` branch
    - Requires a pull request for making changes. There's are requirements for needing at least 1 approval and having status checks pass, but these are bypassable for maintainers (which is sometimes necessary due to social constrainsts 😄 ). Bypassing requires the bypasser to explicitly acknowledge that they are bypassing these restrictions.

2. [RC](https://github.com/digidem/release-automations-expo/settings/rules/3531754)

    - Enforces linear history when pushing commits to `rc/` branches
    - Only allows the Release Bot to create and delete `rc/` branches
    - Prevents force pushes on `rc/` branches
    
3. [Release](https://github.com/digidem/comapeo-mobile/settings/rules/3635895)

    - Only allows the Release Bot to create and delete `release/` branches
    - Requires pull requests to add changes to `release/` branches
        - Requires status checks to pass on PRs before merging
        - Requires a PR to have at least 1 approval in order to merge
        - Only allows using the `Merge` option when merging PRs into `release/` branches
    - Prevents force pushes on `release/` branches
    
4. [Release tags (no update or delete)](https://github.com/digidem/comapeo-mobile/settings/rules/3635901)

    - Prevents anyone from updating or deleting release tags

5. [Release tags (restrict creation)](https://github.com/digidem/comapeo-mobile/settings/rules/3635904)

    - Only allows Release Bot to create release tags

achou11 committed:
define automation workflows

achou11 committed:
remove old release candidate workflow

gmaclennan reviewed (changes requested):
Looking good to me, very happy to have these automations in place and the associated rules, which should significantly reduce the chance of human error during the release process. A couple of notes:

- I think we have landed on the term "scheduled release" instead of "feature release", so we should update it here.
- It would be good to document these different workflows - there are a lot of them now and it's not obvious what they all do and why. There is documentation on the original https://github.com/digidem/release-automations README.md - we could either put a readme here, or we could add comments at the top of each workflow (personally I think I prefer code comments for documentation).
- I think this PR should include updates to the contributor or release process docs for this repo explaining how to do each task like the recipes in https://www.notion.so/digidem/2024-09-06-RFC-release-strategy-85120788688641978fd4cb939c2fcbe5?pvs=4 - this should live in this repo somewhere.

achou11 committed:
rename feature release to scheduled release

achou11 committed:
adjust fallback for comment to trigger rc build

achou11 committed:
add documentation for relevant release workflows

achou11 committed:
add some release docs

achou11 commented:
> I think we have landed on the term "scheduled release" instead of "feature release", so we should update it here.

addressed  via https://github.com/digidem/comapeo-mobile/pull/959/commits/f6a71264b9d0a1c9e550a34ee3e4f15d43dfd924

> It would be good to document these different workflows - there are a lot of them now and it's not obvious what they all do and why. There is documentation on the original https://github.com/digidem/release-automations README.md - we could either put a readme here, or we could add comments at the top of each workflow (personally I think I prefer code comments for documentation).

addressed via https://github.com/digidem/comapeo-mobile/pull/959/commits/6da94b5cf83aaf3f43e5aa6b005188640491b58e
 
> I think this PR should include updates to the contributor or release process docs for this repo explaining how to do each task like the recipes in https://www.notion.so/digidem/2024-09-06-RFC-release-strategy-85120788688641978fd4cb939c2fcbe5?pvs=4 - this should live in this repo somewhere.

addressed via https://github.com/digidem/comapeo-mobile/pull/959/commits/6bb4477eccc5040ec1ebc9ca56036e2c0362e1b6. ~Wasn't sure about where they should live so for now, I created an entirely new document that has some overlap with `BuildAndPublishingFlows.md`. I think the goal would be to replace that existing doc with the new one, once we actually start using these workflows~ EDIT: decided to just remove the old docs, see 35ab304. can still refer to them using the git history if needed

gmaclennan reviewed (approved):
all looks good to me.

achou11 commented on .github/workflows/create-release-candidate.yml:
@ErikSin is it worth keeping this workflow around for a bit longer? Don't want to remove it completely if you may still need it for the current release-in-progress

ErikSin replied:
No, you can delete it now

achou11 committed:
remove old release docs

achou11 committed:
Merge branch 'develop' into 900/release-automations
